File: src/codec/types/structure.cpp
/// StructureSettings codec implementation.

#include "endweave/codec/types/structure.h"

namespace endweave {

namespace {

template <typename PosTag>
std::expected<StructureSettings, ReadError> read_settings(PacketReader& reader) {
    StructureSettings s;
    auto palette = reader.read_string();
    if (!palette) return std::unexpected(palette.error());
    s.palette_name = std::move(*palette);

    auto ignore_ent = reader.read_bool();
    if (!ignore_ent) return std::unexpected(ignore_ent.error());
    s.ignore_entities = *ignore_ent;

    auto ignore_blk = reader.read_bool();
    if (!ignore_blk) return std::unexpected(ignore_blk.error());
    s.ignore_blocks = *ignore_blk;

    auto allow_ntc = reader.read_bool();
    if (!allow_ntc) return std::unexpected(allow_ntc.error());
    s.allow_non_ticking_chunks = *allow_ntc;

    auto size = reader.read<PosTag>();
    if (!size) return std::unexpected(size.error());
    s.size = *size;

    auto offset = reader.read<PosTag>();
    if (!offset) return std::unexpected(offset.error());
    s.offset = *offset;

    auto player = reader.read_varint64();
    if (!player) return std::unexpected(player.error());
    s.last_edit_player = *player;

    auto rot = reader.read_byte();
    if (!rot) return std::unexpected(rot.error());
    s.rotation = *rot;

    auto mir = reader.read_byte();
    if (!mir) return std::unexpected(mir.error());
    s.mirror = *mir;

    auto anim_mode = reader.read_byte();
    if (!anim_mode) return std::unexpected(anim_mode.error());
    s.animation_mode = *anim_mode;

    auto anim_sec = reader.read_float_le();
    if (!anim_sec) return std::unexpected(anim_sec.error());
    s.animation_seconds = *anim_sec;

    auto integrity = reader.read_float_le();
    if (!integrity) return std::unexpected(integrity.error());
    s.integrity_value = *integrity;

    auto seed = reader.read_uint_le();
    if (!seed) return std::unexpected(seed.error());
    s.integrity_seed = *seed;

    auto pivot = reader.read<vec3>();
    if (!pivot) return std::unexpected(pivot.error());
    s.pivot = *pivot;

    return s;
}

template <typename PosTag>
void write_settings(PacketWriter& writer, const StructureSettings& s) {
    writer.write_string(s.palette_name);
    writer.write_bool(s.ignore_entities);
    writer.write_bool(s.ignore_blocks);
    writer.write_bool(s.allow_non_ticking_chunks);
    writer.write<PosTag>(s.size);
    writer.write<PosTag>(s.offset);
    writer.write_varint64(s.last_edit_player);
    writer.write_byte(s.rotation);
    writer.write_byte(s.mirror);
    writer.write_byte(s.animation_mode);
    writer.write_float_le(s.animation_seconds);
    writer.write_float_le(s.integrity_value);
    writer.write_uint_le(s.integrity_seed);
    writer.write<vec3>(s.pivot);
}

} // namespace

template <> auto PacketReader::read<structure_settings_v924>() -> std::expected<StructureSettings, ReadError> {
    return read_settings<network_block_pos>(*this);
}
template <> auto PacketReader::read<structure_settings_v944>() -> std::expected<StructureSettings, ReadError> {
    return read_settings<block_pos>(*this);
}
template <> void PacketWriter::write<structure_settings_v924>(const StructureSettings& val) {
    write_settings<network_block_pos>(*this, val);
}
template <> void PacketWriter::write<structure_settings_v944>(const StructureSettings& val) {
    write_settings<block_pos>(*this, val);
}

} // namespace endweave

File: src/endstone_endweave/codec/types/structure.py
"""Structure-related compound types (StructureSettings)."""

from dataclasses import dataclass

from endstone_endweave.codec.reader import PacketReader
from endstone_endweave.codec.writer import PacketWriter

from .primitives import (
    BLOCK_POS,
    BOOL,
    BYTE,
    FLOAT_LE,
    NETWORK_BLOCK_POS,
    STRING,
    UINT_LE,
    VAR_INT64,
    VEC3,
    Type,
)


@dataclass
class StructureSettings:
    """StructureSettings used by StructureBlockUpdate and StructureTemplateDataRequest.

    Attributes:
        palette_name: Structure palette name.
        ignore_entities: Whether to ignore entities.
        ignore_blocks: Whether to ignore blocks.
        allow_non_ticking_chunks: Whether to allow non-ticking chunks.
        size: Structure size as (x, y, z).
        offset: Structure offset as (x, y, z).
        last_edit_player: Unique ID of last editing player.
        rotation: Rotation enum value.
        mirror: Mirror enum value.
        animation_mode: Animation mode enum value.
        animation_seconds: Animation duration in seconds.
        integrity_value: Structure integrity (0.0 - 100.0).
        integrity_seed: Integrity seed.
        pivot: Rotation pivot point (x, y, z).
    """

    palette_name: str
    ignore_entities: bool
    ignore_blocks: bool
    allow_non_ticking_chunks: bool
    size: tuple[int, int, int]
    offset: tuple[int, int, int]
    last_edit_player: int
    rotation: int
    mirror: int
    animation_mode: int
    animation_seconds: float
    integrity_value: float
    integrity_seed: int
    pivot: tuple[float, float, float]


class _StructureSettingsV924Type(Type["StructureSettings"]):
    """v924 StructureSettings: Size/Offset encoded as NetworkBlockPos (uvarint Y)."""

    _pos_type: Type[tuple[int, int, int]] = NETWORK_BLOCK_POS

    def read(self, reader: PacketReader) -> StructureSettings:
        return StructureSettings(
            palette_name=STRING.read(reader),
            ignore_entities=BOOL.read(reader),
            ignore_blocks=BOOL.read(reader),
            allow_non_ticking_chunks=BOOL.read(reader),
            size=self._pos_type.read(reader),
            offset=self._pos_type.read(reader),
            last_edit_player=VAR_INT64.read(reader),
            rotation=BYTE.read(reader),
            mirror=BYTE.read(reader),
            animation_mode=BYTE.read(reader),
            animation_seconds=FLOAT_LE.read(reader),
            integrity_value=FLOAT_LE.read(reader),
            integrity_seed=UINT_LE.read(reader),
            pivot=VEC3.read(reader),
        )

    def write(self, writer: PacketWriter, value: StructureSettings) -> None:
        STRING.write(writer, value.palette_name)
        BOOL.write(writer, value.ignore_entities)
        BOOL.write(writer, value.ignore_blocks)
        BOOL.write(writer, value.allow_non_ticking_chunks)
        self._pos_type.write(writer, value.size)
        self._pos_type.write(writer, value.offset)
        VAR_INT64.write(writer, value.last_edit_player)
        BYTE.write(writer, value.rotation)
        BYTE.write(writer, value.mirror)
        BYTE.write(writer, value.animation_mode)
        FLOAT_LE.write(writer, value.animation_seconds)
        FLOAT_LE.write(writer, value.integrity_value)
        UINT_LE.write(writer, value.integrity_seed)
        VEC3.write(writer, value.pivot)


class _StructureSettingsV944Type(_StructureSettingsV924Type):
    """v944 StructureSettings: Size/Offset encoded as BlockPos (varint Y)."""

    _pos_type: Type[tuple[int, int, int]] = BLOCK_POS


STRUCTURE_SETTINGS_V924 = _StructureSettingsV924Type()
STRUCTURE_SETTINGS_V944 = _StructureSettingsV944Type()
